pandas/core/arrays/floating.py

```diff
@@ -0,0 +1,641 @@
+        # and mask the result.
+        if method == "reduce":
+            # Not clear how to handle missing values in reductions. Raise.
+            raise NotImplementedError("The 'reduce' method is not supported.")
+        out = kwargs.get("out", ())
+
+        for x in inputs + out:
+            if not isinstance(x, self._HANDLED_TYPES + (FloatingArray,)):
+                return NotImplemented
+
+        # for binary ops, use our custom dunder methods
+        result = ops.maybe_dispatch_ufunc_to_dunder_op(
+            self, ufunc, method, *inputs, **kwargs
+        )
+        if result is not NotImplemented:
+            return result
+
+        mask = np.zeros(len(self), dtype=bool)
+        inputs2 = []
+        for x in inputs:
+            if isinstance(x, FloatingArray):
+                mask |= x._mask
+                inputs2.append(x._data)
+            else:
+                inputs2.append(x)
+
+        def reconstruct(x):
+            # we don't worry about scalar `x` here, since we
+            # raise for reduce up above.
+
+            # TODO
+            if is_float_dtype(x.dtype):
+                m = mask.copy()
+                return FloatingArray(x, m)
+            else:
+                x[mask] = np.nan
+            return x
+
+        result = getattr(ufunc, method)(*inputs2, **kwargs)
+        if isinstance(result, tuple):
+            tuple(reconstruct(x) for x in result)
+        else:
+            return reconstruct(result)
+
+    def _coerce_to_array(self, value) -> Tuple[np.ndarray, np.ndarray]:
+        return coerce_to_array(value, dtype=self.dtype)
+
+    def astype(self, dtype, copy: bool = True) -> ArrayLike:
+        """
+        Cast to a NumPy array or ExtensionArray with 'dtype'.
+
+        Parameters
+        ----------
+        dtype : str or dtype
+            Typecode or data-type to which the array is cast.
+        copy : bool, default True
+            Whether to copy the data, even if not necessary. If False,
+            a copy is made only if the old dtype does not match the
+            new dtype.
+
+        Returns
+        -------
+        ndarray or ExtensionArray
+            NumPy ndarray, or BooleanArray, IntegerArray or FloatingArray with
+            'dtype' for its dtype.
+
+        Raises
+        ------
+        TypeError
+            if incompatible type with an FloatingDtype, equivalent of same_kind
+            casting
+        """
+        from pandas.core.arrays.boolean import BooleanArray, BooleanDtype
+        from pandas.core.arrays.integer import _IntegerDtype, IntegerArray
+        from pandas.core.arrays.string_ import StringDtype, StringArray
+
+        dtype = pandas_dtype(dtype)
+
+        # if we are astyping to an existing FloatingDtype we can fastpath
+        if isinstance(dtype, _FloatingDtype):
+            result = self._data.astype(dtype.numpy_dtype, copy=False)
+            return type(self)(result, mask=self._mask, copy=False)
```

Comment:
Why are these both `copy=False`? Shouldn't the first be `copy=copy`?

Reply:
Fix for the integer case -> https://github.com/pandas-dev/pandas/pull/34931 (can copy that over here once the other is merged)